### DISCUSSION TITLE
How to fail validation from a specific validator?

### DISCUSSION THREAD

**Discussion by chbndrhnns** [pydantic/pydantic#4914] [Question]
Hi,

I have this snippet which I expect to fail in `_enforce_type`. However, it fails in `_generate`:

```
cls = <class 'tests.test_.M'>, val = None, values = {}

    @validator("description", always=True)
    def _generate(cls, val, values):
>       return values["url"].path
E       KeyError: 'url'
```

The union on `url` exists to see no warning in PyCharm when passing in a string instead of constructing an instance of `AnyHttpUrl`.

Now I am wondering if there is a way to make this example work. I'd like to fail validation early in `_enforce_type`.
Any ideas?

```py
from typing import Optional, Union

from pydantic import AnyHttpUrl, BaseModel, parse_obj_as, validator


class M(BaseModel):
    url: Union[str, AnyHttpUrl]
    description: Optional[str]

    @validator("url", pre=True)
    def _enforce_type(cls, val):
        return parse_obj_as(AnyHttpUrl, val)

    @validator("description", always=True)
    def _generate(cls, val, values):
        return values["url"].path


def test_():
    actual = M(url="/abc")
    assert actual.description == "/abc"
```

**Comment by samuelcolvin** [comment]
Don't use a union, that's wrong. Just mute the mypy warning.

**Reply by chbndrhnns** [reply]
Understood. 
However, I am not using mypy but the pydantic PyCharm plugin which triggers the warnings. It seems it does not understand that `AnyHttpUrl` inherits from str and thus passing a string should be ok.
The implication would be to add lots of `# noinspection PyTypeChecker` comments.

mypy does not complain about passing in a string.

**Reply by chbndrhnns** [reply]
Added an issue in the project: https://github.com/koxudaxi/pydantic-pycharm-plugin/issues/624

**Reply by samuelcolvin** [reply]
This isn't new, best to check for duplicates before creating new issues. Should be fixed in V2, see #4680.
